FF ONLY Merge 0.6.x into master, September 5

sjrd opened the pull request:
```
$ git checkout -b merge-0.6.x-into-master-september-5
Switched to a new branch 'merge-0.6.x-into-master-september-5'
```
```
$ export mb=$(git merge-base scalajs/0.6.x scalajs/master)
```
```
$ git log --graph --oneline --decorate $mb..scalajs/0.6.x | cat
*   c72bf677d (scalajs/0.6.x, origin/0.6.x) Merge pull request #3742 from sjrd/no-javaconverters-in-test-suite
|\  
| * 934ba738b Rewrite the collection behind `AbstractCollectionTest` not to use `Box`.
| * 61972891e Remove two unused imports of `JavaConverters`.
| * 6fbe7b2ba Do not rely on JavaConverters in the test suite.
|/  
*   1f3afeb8e Merge pull request #3690 from exoego/uri
|\  
| * f0a0228eb Various fixes in URI
|/  
*   a3877296b Merge pull request #3757 from sjrd/fix-incremental-compilation-in-sbt-1
|\  
| * 78682118e Fix #3419 and #3596: Backup and restore .sjsir files ourselves.
| * 649bf36ae Upgrade the sbt plugin to sbt 1.2.1.
* |   f3853d6be Merge pull request #3736 from sjrd/direct-implem-of-priority-queue
|\ \  
| |/  
|/|   
| * 67d32b005 Implement `java.util.PriorityQueue` from scratch.
* |   34267db42 Merge pull request #3759 from sjrd/disable-slow-tests-in-rhino
|\ \  
| * | eaa8d543c [no-master] Disable slow scalaTestSuite tests when running in Rhino.
|/ /  
* |   96cefda69 Merge pull request #3753 from sjrd/fast-isinstanceof-0.6.x
|\ \  
| |/  
|/|   
| * 108984703 [no-master] Fix #2066: Emit `x.isInstanceOf[C]` as `x instanceof C` when possible.
|/  
*   a347ca108 Merge pull request #3748 from sjrd/java-lang-nums-decode
|\  
| * 03a9acdce (origin/java-lang-nums-decode) Implement `java.lang.{Byte,Short,Integer.Long}.decode()`.
* |   ae0038221 Merge pull request #3755 from sjrd/hotfix-unicode-database-jdk8u211
|\ \  
| |/  
|/|   
| * b7203dda3 (origin/hotfix-unicode-database-jdk8u211) Update the Unicode database to match updates from JDK 8u211.
|/  
*   f28eab671 Merge pull request #3746 from sjrd/remove-dead-code-allfactories
|\  
| * 434b8ce12 Remove dead-code definitions of `allFactories` in the test suite.
|/  
*   a77a0dda2 Merge pull request #3744 from exoego/linking-exception-message
|\  
| * 783c6bec6 Suppress stack trace of LinkingException's in sbt plugin since it is useless for users
|/  
*   0428cd815 Merge pull request #3724 from exoego/module-initializer-message
|\  
| * 03e21895f Fix #3697: Improve error message when no main module initializer was found
* |   1828e5b51 Merge pull request #3741 from sjrd/more-tests-for-ju-map
|\ \  
| |/  
|/|   
| * a5c1d2095 Enhance the test suite for `java.util.Map`s.
| * d0d04a5a8 Remove dead code in the javalib test suites.
|/  
*   14d5dc1c0 Merge pull request #3735 from sjrd/implement-concurrent-hashmap-with-ju-hashmap
|\  
| * 60c94e28c (origin/implement-concurrent-hashmap-with-ju-hashmap) Implement `j.u.c.ConcurrentHashMap` on top of `j.u.HashMap`.
|/  
*   710bf7c54 Merge pull request #3733 from sjrd/use-objects-equals-hashcode
|\  
| * e07d99d3e (origin/use-objects-equals-hashcode) Use `Objects.equals()` and `Objects.hashCode()` where relevant.
|/  
* 762243810 Merge pull request #3708 from sjrd/direct-implem-of-jdk-hash-maps
* dc54bf15f (origin/direct-implem-of-jdk-hash-maps) Implement `j.u.{Linked,}HashSet` on top of `j.u.{Linked,}HashMap`.
* 2f519823e Implement java.util.{Linked,}HashMap from scratch.
* 417c3ed30 Implement j.u.Hashtable on top of j.u.HashMap.
```
```
$ git merge --no-commit a347ca108
Auto-merging sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala
CONFLICT (content): Merge conflict in sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala
Auto-merging javalib/src/main/scala/java/util/package.scala
CONFLICT (content): Merge conflict in javalib/src/main/scala/java/util/package.scala
Auto-merging javalib/src/main/scala/java/util/Collections.scala
Auto-merging javalanglib/src/main/scala/java/lang/Short.scala
Auto-merging javalanglib/src/main/scala/java/lang/Long.scala
Auto-merging javalanglib/src/main/scala/java/lang/Integer.scala
Auto-merging javalanglib/src/main/scala/java/lang/Character.scala
Auto-merging javalanglib/src/main/scala/java/lang/Byte.scala
Recorded preimage for 'javalib/src/main/scala/java/util/package.scala'
Recorded preimage for 'sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala'
Automatic merge failed; fix conflicts and then commit the result.
```
```
$ git st
M  javalanglib/src/main/scala/java/lang/Byte.scala
M  javalanglib/src/main/scala/java/lang/Character.scala
M  javalanglib/src/main/scala/java/lang/Integer.scala
M  javalanglib/src/main/scala/java/lang/Long.scala
M  javalanglib/src/main/scala/java/lang/Short.scala
M  javalib/src/main/scala/java/util/AbstractCollection.scala
M  javalib/src/main/scala/java/util/AbstractList.scala
M  javalib/src/main/scala/java/util/AbstractMap.scala
M  javalib/src/main/scala/java/util/ArrayDeque.scala
M  javalib/src/main/scala/java/util/Arrays.scala
M  javalib/src/main/scala/java/util/Collections.scala
M  javalib/src/main/scala/java/util/HashMap.scala
M  javalib/src/main/scala/java/util/HashSet.scala
M  javalib/src/main/scala/java/util/Hashtable.scala
M  javalib/src/main/scala/java/util/LinkedHashMap.scala
M  javalib/src/main/scala/java/util/LinkedHashSet.scala
M  javalib/src/main/scala/java/util/LinkedList.scala
A  javalib/src/main/scala/java/util/NullRejectingHashMap.scala
M  javalib/src/main/scala/java/util/Objects.scala
M  javalib/src/main/scala/java/util/PriorityQueue.scala
M  javalib/src/main/scala/java/util/concurrent/ConcurrentHashMap.scala
M  javalib/src/main/scala/java/util/concurrent/CopyOnWriteArrayList.scala
UU javalib/src/main/scala/java/util/package.scala
UU sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/ByteTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/IntegerTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/LongTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/ShortTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/AbstractSetTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/ArrayDequeTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/CollectionTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/DequeTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/HashMapTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/HashSetTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/LinkedHashMapTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/LinkedHashSetTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/ListTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/MapTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/NavigableSetTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/SetTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/SortedMapTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/SortedSetTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/TreeSetTest.scala
A  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/TrivialImmutableMap.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/ConcurrentHashMapTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/ConcurrentMapTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/ConcurrentSkipListSetTest.scala
```
[...] Fix conflicts, type errors and javalanglib IR patching errors
```diff
$ git diff | cat
diff --cc javalib/src/main/scala/java/util/package.scala
index f1d77abd8,4f87826ba..000000000
--- a/javalib/src/main/scala/java/util/package.scala
+++ b/javalib/src/main/scala/java/util/package.scala
diff --cc sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala
index b9b97db5a,f4b24648f..000000000
--- a/sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala
+++ b/sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala
@@@ -223,16 -322,13 +223,21 @@@ private[sbtplugin] object ScalaJSPlugin
  
              IO.createDirectory(output.getParentFile)
  
 +            def relURI(path: String) = new URI(null, null, path, null)
 +
 +            val out = LinkerOutput(LinkerOutput.newAtomicPathFile(output.toPath))
 +              .withSourceMap(LinkerOutput.newAtomicPathFile(sourceMapFile.toPath))
 +              .withSourceMapURI(relURI(sourceMapFile.getName))
 +              .withJSFileURI(relURI(output.getName))
 +
-             enhanceIRVersionNotSupportedException {
-               val tlog = sbtLogger2ToolsLogger(log)
-               await(log)(linker.link(ir, moduleInitializers, out, tlog)(_))
+             try {
 -              linker.link(ir, moduleInitializers,
 -                  AtomicWritableFileVirtualJSFile(output),
 -                  sbtLogger2ToolsLogger(log))
++              enhanceIRVersionNotSupportedException {
++                val tlog = sbtLogger2ToolsLogger(log)
++                await(log)(linker.link(ir, moduleInitializers, out, tlog)(_))
++              }
+             } catch {
+               case e: LinkingException =>
+                 throw new MessageOnlyException(e.getMessage)
              }
  
              logIRCacheStats(log)
@@@ -387,15 -914,23 +392,24 @@@
          (scalaJSModuleInitializers in (This, Zero, This)).value,
  
        scalaJSModuleInitializers ++= {
 -        val mainClasses = scalaJSDiscoveredMainClasses.value
++        val mainClasses = discoveredMainClasses.value
          if (scalaJSUseMainModuleInitializer.value) {
            Seq(scalaJSMainModuleInitializer.value.getOrElse {
-             throw new MessageOnlyException(
-                 "No main module initializer was specified (possibly because " +
-                 "no or multiple main classes were found), but " +
-                 "scalaJSUseMainModuleInitializer was set to true. " +
-                 "You can explicitly specify it either with " +
-                 "`mainClass := Some(...)` or with " +
-                 "`scalaJSMainModuleInitializer := Some(...)`")
+             if (mainClasses.isEmpty) {
+               throw new MessageOnlyException(
+                   "No main module initializer was specified, but " +
+                   "scalaJSUseMainModuleInitializer was set to true. " +
+                   "You can explicitly specify it either with " +
+                   "`mainClass := Some(...)` or with " +
+                   "`scalaJSMainModuleInitializer := Some(...)`")
+             } else {
+               throw new MessageOnlyException(
 -                  s"Multiple main classes (${mainClasses.keys}) were found. " +
++                  s"Multiple main classes (${mainClasses.mkString(", ")}) " +
++                  "were found. " +
+                   "You can explicitly specify the one you want with " +
+                   "`mainClass := Some(...)` or with " +
+                   "`scalaJSMainModuleInitializer := Some(...)`")
+             }
            })
          } else {
            Seq.empty
diff --git a/javalanglib/src/main/scala/java/lang/Integer.scala b/javalanglib/src/main/scala/java/lang/Integer.scala
index 94bf02dfe..3c9cb5dac 100644
--- a/javalanglib/src/main/scala/java/lang/Integer.scala
+++ b/javalanglib/src/main/scala/java/lang/Integer.scala
@@ -123,7 +123,7 @@ object Integer {
     decodeGeneric(nm, valueOf(_, _))
 
   @inline private[lang] def decodeGeneric[A](nm: String,
-      parse: (String, Int) => A): A = {
+      parse: js.Function2[String, Int, A]): A = {
 
     val len = nm.length()
     var i = 0
```
```
$ git add -u
```
```
$ git commit
Recorded resolution for 'javalib/src/main/scala/java/util/package.scala'.
Recorded resolution for 'sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala'.
[merge-0.6.x-into-master-september-5 5dc8e67e6] Merge '0.6.x' into 'master'.
```
```
$ git merge -s ours 34267db42
Merge made by the 'ours' strategy.
```
```
$ git show
commit 62d352ec27712d19ba63f2ffd87c6e720990daf8 (HEAD -> merge-0.6.x-into-master-september-5)
Merge: 5dc8e67e6 34267db42
Author: Sébastien Doeraene <sjrdoeraene@gmail.com>
Date:   Thu Sep 5 18:28:17 2019 +0200

    Skip the [no-master] commits 108984703 and eaa8d543c.

```
```
$ git merge --no-commit scalajs/0.6.x 
CONFLICT (modify/delete): test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/package.scala deleted in scalajs/0.6.x and modified in HEAD. Version HEAD of test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/package.scala left in tree.
Auto-merging scripts/publish.sh
CONFLICT (content): Merge conflict in scripts/publish.sh
CONFLICT (modify/delete): sbt-plugin/src/main/scala-sbt-1.0/org/scalajs/sbtplugin/SBTCompat.scala deleted in HEAD and modified in scalajs/0.6.x. Version scalajs/0.6.x of sbt-plugin/src/main/scala-sbt-1.0/org/scalajs/sbtplugin/SBTCompat.scala left in tree.
Auto-merging project/Build.scala
CONFLICT (content): Merge conflict in project/Build.scala
CONFLICT (modify/delete): js-envs/src/main/scala/org/scalajs/jsenv/nodejs/AbstractNodeJSEnv.scala deleted in HEAD and modified in scalajs/0.6.x. Version scalajs/0.6.x of js-envs/src/main/scala/org/scalajs/jsenv/nodejs/AbstractNodeJSEnv.scala left in tree.
CONFLICT (modify/delete): js-envs/src/main/scala/org/scalajs/jsenv/ExternalJSEnv.scala deleted in HEAD and modified in scalajs/0.6.x. Version scalajs/0.6.x of js-envs/src/main/scala/org/scalajs/jsenv/ExternalJSEnv.scala left in tree.
Auto-merging javalib/src/main/scala/java/net/URI.scala
Recorded preimage for 'project/Build.scala'
Recorded preimage for 'scripts/publish.sh'
Automatic merge failed; fix conflicts and then commit the result.
```
[...] Fix conflicts
```
$ git st
M  javalib/src/main/scala/java/net/URI.scala
A  javalib/src/main/scala/java/util/NaturalComparator.scala
M  javalib/src/main/scala/java/util/PriorityQueue.scala
DU js-envs/src/main/scala/org/scalajs/jsenv/ExternalJSEnv.scala
DU js-envs/src/main/scala/org/scalajs/jsenv/nodejs/AbstractNodeJSEnv.scala
UU project/Build.scala
M  sbt-plugin-test/project/build.properties
DU sbt-plugin/src/main/scala-sbt-1.0/org/scalajs/sbtplugin/SBTCompat.scala
A  sbt-plugin/src/main/scala-sbt-1.0/org/scalajs/sbtplugin/SJSIRFileManager.scala
UU scripts/publish.sh
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/net/URITest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/AbstractCollectionTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/ArrayDequeTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/CollectionTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/CollectionsOnCheckedCollectionTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/CollectionsOnCheckedListTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/CollectionsOnCheckedMapTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/CollectionsOnCollectionsTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/CollectionsOnListsTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/CollectionsOnSetsTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/CollectionsTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/HashtableTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/LinkedHashMapTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/LinkedHashSetTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/LinkedListTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/ListTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/MapTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/NavigableSetTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/PriorityQueueTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/PropertiesTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/SetTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/SortedSetTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/TreeSetTest.scala
A  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/TrivialImmutableCollection.scala
A  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/Utils.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/ConcurrentLinkedQueueTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/ConcurrentSkipListSetTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/concurrent/CopyOnWriteArrayListTest.scala
UD test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/package.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/niocharset/CharsetTest.scala
M  test-suite/shared/src/test/scala/org/scalajs/testsuite/utils/CollectionsTestBase.scala
```
```
$ git rm -f test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/package.scala
test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/package.scala: needs merge
rm 'test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/package.scala'
```
```
$ git rm -f js-envs/src/main/scala/org/scalajs/jsenv/ExternalJSEnv.scala
js-envs/src/main/scala/org/scalajs/jsenv/ExternalJSEnv.scala: needs merge
rm 'js-envs/src/main/scala/org/scalajs/jsenv/ExternalJSEnv.scala'
```
```
$ git rm -f js-envs/src/main/scala/org/scalajs/jsenv/nodejs/AbstractNodeJSEnv.scala
js-envs/src/main/scala/org/scalajs/jsenv/nodejs/AbstractNodeJSEnv.scala: needs merge
rm 'js-envs/src/main/scala/org/scalajs/jsenv/nodejs/AbstractNodeJSEnv.scala'
```
```diff
$ git diff | cat
diff --cc project/Build.scala
index e23a7a3b8,40419ca22..000000000
--- a/project/Build.scala
+++ b/project/Build.scala
diff --cc scripts/publish.sh
index e2c750d83,2b075bea8..000000000
--- a/scripts/publish.sh
+++ b/scripts/publish.sh
* Unmerged path sbt-plugin/src/main/scala-sbt-1.0/org/scalajs/sbtplugin/SBTCompat.scala
```
```
$ git add -u
```
```
$ git commit
Recorded resolution for 'project/Build.scala'.
Recorded resolution for 'scripts/publish.sh'.
[merge-0.6.x-into-master-september-5 fd5e9a2b2] Merge '0.6.x' into 'master'.
```

gzm0 commented on sbt-plugin/src/main/scala-sbt-1.0/org/scalajs/sbtplugin/SBTCompat.scala:
```diff
@@ -0,0 +1,53 @@
+      incOptions.externalHooks.withExternalClassFileManager(sjsirFileManager)
+    incOptions.withExternalHooks(newExternalHooks)
+  }
+}
```
This should go into a non-version specific directory in master, no?

gzm0 commented on sbt-plugin/src/main/scala-sbt-1.0/org/scalajs/sbtplugin/SJSIRFileManager.scala:
```diff
@@ -0,0 +1,117 @@
+    IO.move(c, target)
+    target
+  }
+}
```
Same here

sjrd commented:
Ouch, well spotted! I wonder why I didn't get a conflict at least on `SBTCompat`...

Fixed with:
```
$ git mv sbt-plugin/src/main/scala-sbt-1.0/org/scalajs/sbtplugin/SJSIRFileManager.scala sbt-plugin/src/main/scala/org/scalajs/sbtplugin/
```
```
$ git rm sbt-plugin/src/main/scala-sbt-1.0/org/scalajs/sbtplugin/SBTCompat.scala
rm 'sbt-plugin/src/main/scala-sbt-1.0/org/scalajs/sbtplugin/SBTCompat.scala'
```
```
$ git st
D  sbt-plugin/src/main/scala-sbt-1.0/org/scalajs/sbtplugin/SBTCompat.scala
R  sbt-plugin/src/main/scala-sbt-1.0/org/scalajs/sbtplugin/SJSIRFileManager.scala -> sbt-plugin/src/main/scala/org/scalajs/sbtplugin/SJSIRFileManager.scala
 M sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala
```
[...] Edit `ScalaJSPluginInternal.scala` to include the changes that were done in `SBTCompat.scala`.
```diff
$ git diff | cat
diff --git a/sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala b/sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala
index 719312112..999c8ded4 100644
--- a/sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala
+++ b/sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala
@@ -111,36 +111,11 @@ private[sbtplugin] object ScalaJSPluginInternal {
     }
   }

-  /** Patches the IncOptions so that .sjsir files are pruned as needed.
-   *
-   *  This complicated logic patches the ClassfileManager factory of the given
-   *  IncOptions with one that is aware of .sjsir files emitted by the Scala.js
-   *  compiler. This makes sure that, when a .class file must be deleted, the
-   *  corresponding .sjsir file are also deleted.
+  /** Patches the IncOptions so that .sjsir files are pruned, backed up and
+   *  restored as needed.
    */
   def scalaJSPatchIncOptions(incOptions: IncOptions): IncOptions = {
-    import xsbti.compile.{ClassFileManager, ClassFileManagerUtil}
-
-    val sjsirFileManager = new ClassFileManager {
-      private[this] val inherited =
-        ClassFileManagerUtil.getDefaultClassFileManager(incOptions)
-
-      def delete(classes: Array[File]): Unit = {
-        inherited.delete(classes.flatMap { classFile =>
-          if (classFile.getPath.endsWith(".class")) {
-            val f = new File(classFile.getPath.stripSuffix(".class") + ".sjsir")
-            if (f.exists) List(f)
-            else Nil
-          } else {
-            Nil
-          }
-        })
-      }
-
-      def generated(classes: Array[File]): Unit = {}
-      def complete(success: Boolean): Unit = {}
-    }
-
+    val sjsirFileManager = new SJSIRFileManager
     val newExternalHooks =
       incOptions.externalHooks.withExternalClassFileManager(sjsirFileManager)
     incOptions.withExternalHooks(newExternalHooks)
```
```
$ git add -u
```
```
$ git amend
```
